cmd/config.go

```diff
@@ -254,9 +259,78 @@ func validateConfig(conf Config, isExecutable func(string) bool) error {
 		}
 	}
 
+	// If there are thresholds to validate, the registry paramater is not allowed to be nil.
+	// Note that the reason for passing it as a pointer in the first place is
+	// because it holds a Mutex, which effectively forbids passing it by value.
+	if conf.Thresholds != nil && len(conf.Thresholds) > 0 && registry == nil {
+		err := fmt.Errorf(
+			"unable to validate thresholds configuration; " +
+				"reason: provided registry is nil",
+		)
+		errList = append(errList, err)
+		return consolidateErrorMessage(errList, "there were problems while validating the specified script configuration: ")
+	}
+
+	for thresholdName, thresholds := range conf.Thresholds {
+		// Fetch metric matching the threshold's name
+		metric, ok := registry.Get(thresholdName)
+		if !ok {
+			// The defined threshold applies to a non-existing metrics
+			err := fmt.Errorf("invalid threshold defined on %s; reason: no metric named %s found", thresholdName, thresholdName)
+			return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+		}
+
+		// Validate the threshold definition against its matching
+		// metric.
+		err := validateThresholdsConfig(thresholdName, thresholds, metric)
+		if err != nil {
+			return errext.WithExitCodeIfNone(err, exitcodes.InvalidConfig)
+		}
+	}
+
 	return consolidateErrorMessage(errList, "There were problems with the specified script configuration:")
 }
 
+// validateThresholdsConfig validates a threshold definition is consistent with the metric it applies to.
+// Given a threshold name, expressions and a metric to apply the expressions too, validateThresholdConfig will
+// assert that each expression uses an aggregation method that's supported by the provided metric. It returns
+// an error otherwise. Note that this function expects the passed in thresholds to have been parsed already, and
+// have their Parsed (ThresholdExpression) field already filled.
+func validateThresholdsConfig(thresholdName string, expressions stats.Thresholds, metric *stats.Metric) error {
+	var supportedMethods []string
+
+	switch metric.Type {
+	case stats.Counter:
+		supportedMethods = []string{stats.TokenCount, stats.TokenRate}
+	case stats.Gauge:
+		supportedMethods = []string{stats.TokenValue}
+	case stats.Rate:
+		supportedMethods = []string{stats.TokenRate}
+	case stats.Trend:
+		supportedMethods = []string{
+			stats.TokenAvg,
+			stats.TokenMin,
+			stats.TokenMax,
+			stats.TokenMed,
+			stats.TokenPercentile,
```

Comment:
this doesn't really check if the percentile is valid

Reply:
I'm uncertain if I understand why, could you elaborate?

In the current state of the PR, one could argue the validation is done in two phases: parsing, which happens during bundling as of this PR's state (expression format validation), and actual validation which happens in config validation: "does the metric to you apply that expression to supports the operation?".

My initial interpretation of your comment is "this does not verify that for a given 'p({value})' percentile expression, the metric it applies to actually holds samples for the selected percentile's `{value}`" would that be correct?

Reply:
I guess you are correct, I probably commented on this *before* I made the general comment that currently parsing of a threshold will still fail the test.